### PR TITLE
Add bounds checking to WBWIIteratorImpl and respect bounds of ReadOptions in Transaction

### PR DESCRIPTION
**Pull request by ywave620**
Fix https://github.com/facebook/rocksdb/issues/11607
Fix https://github.com/facebook/rocksdb/issues/11679
Fix https://github.com/facebook/rocksdb/issues/11606
Fix https://github.com/facebook/rocksdb/issues/2343

Summary:

Add bounds checking to `WBWIIteratorImpl`, which will be reflected in `BaseDeltaIterator::delta_iterator_::Valid()`, just like `BaseDeltaIterator::base_iterator_::Valid()`. In this way, the two sub itertors become more aligned from `BaseDeltaIterator`'s perspective. Like `DBIter`, the added bounds checking caps in either bound when seeking and disvalidates the `WBWIIteratorImpl` iterator when the lower bound is past or the upper bound is reached.

Test Plan:

- A simple test added to write_batch_with_index_test.cc to exercise the bounds checking in `WBWIIteratorImpl`.
- A sophisticated test added to transaction_test.cc to assert that `Transaction` with different write policies honor bounds in `ReadOptions`. It should be so as long as the  `BaseDeltaIterator` is correctly coordinating the two sub iterators to perform iterating and bounds checking.

### CODE REVIEW

**Comment by ajkr** [comment]
> Add bounds checking to WBWIIteratorImpl, which will be reflected in BaseDeltaIterator::delta_iterator_::Valid()

Interesting timing - https://github.com/facebook/rocksdb/pull/11662#discussion_r1285252595. Does this PR also fix the two issues mentioned in #11662?

**Comment by cz2h** [comment]
> > Add bounds checking to WBWIIteratorImpl, which will be reflected in BaseDeltaIterator::delta_iterator_::Valid()
> 
> Interesting timing - [#11662 (comment)](https://github.com/facebook/rocksdb/pull/11662#discussion_r1285252595). Does this PR also fix the two issues mentioned in #11662?

I think so. This PR also provide feature for lower_bound as well. Looks like multiple issue is pointing to the same fix.

**Comment by ywave620** [comment]
@cz2h Thanks for the information. I agree with that this PR can be a replacer for the one you open, https://github.com/facebook/rocksdb/pull/11662, and could I move the test you wrote here?

**Comment by cz2h** [comment]
> @cz2h Thanks for the information. I agree with that this PR can be a replacer for the one you open, #11662, and could I move the test you wrote here?

Please do, thanks.

File: utilities/write_batch_with_index/write_batch_with_index_internal.h [hunk 1]
```diff
@@ -233,13 +256,24 @@ class WBWIIteratorImpl : public WBWIIterator {
...
 
   void SeekForPrev(const Slice& key) override {
+    if (afterUpperBound(&key)) {
+      WriteBatchIndexEntry search_entry(
+          iterate_upper_bound_, column_family_id_,
+          // is_forward_direction, set to true so that
+          // iterate_upper_bound_ is excluded, this is not a readable trick,
+          // TODO fix it by renaming WriteBatchIndexEntry::is_forward_direction
+          true, false /* is_seek_to_first */);
+      skip_list_iter_.SeekForPrev(&search_entry);
```
Comment (cbi42) [inline utilities/write_batch_with_index/write_batch_with_index_internal.h:273]: It seems the iterator can incorrectly return a key equal to `iterate_upper_bound_`. Maybe we can just do SeekToLast()?

File: utilities/write_batch_with_index/write_batch_with_index_internal.h [hunk 2]
```diff
@@ -228,13 +216,14 @@ class WBWIIteratorImpl : public WBWIIterator {
...
+          nullptr /* search_key */, column_family_id_,
+          true /* is_forward_direction */, true /* is_seek_to_first */);
+      skip_list_iter_.Seek(&search_entry);
     }
 
-    WriteBatchIndexEntry search_entry(
-        nullptr /* search_key */, column_family_id_,
-        true /* is_forward_direction */, true /* is_seek_to_first */);
-    skip_list_iter_.Seek(&search_entry);
+    out_of_bound_ = false;
```
Comment (cbi42) [inline utilities/write_batch_with_index/write_batch_with_index_internal.h:226]: Need to check upper bound too? Other Seek*() calls also miss a similar check.

**Review by cbi42 — CHANGES_REQUESTED** [review]

File: utilities/write_batch_with_index/write_batch_with_index_internal.h (hunk 1 above)
Comment (cbi42) [reply utilities/write_batch_with_index/write_batch_with_index_internal.h:275]: Do you think this can happen?

**Comment by cbi42** [comment]
> @cbi42 Any plan to move this forward? (BTW, the CI failed on installing dependencies)

Hi - sorry for the delay. I plan to finish reviewing this week.

File: utilities/write_batch_with_index/write_batch_with_index_internal.h (hunk 1 above)
Comment (cbi42) [reply utilities/write_batch_with_index/write_batch_with_index_internal.h:275]: From https://github.com/facebook/rocksdb/blob/d98a9cfb27dbffb4f7a7ab173b366dd626c2f721/db/write_batch.cc#L11-L12, it seems entries won't have offset = 0. However, I feel it's safer not to assume this, and just call SeekToLast().

**Comment by cbi42** [comment]
> @cbi42 Any plan to move this forward? (BTW, the CI failed on installing dependencies)

@ywave620 There are a few comments that are still not addressed, can you take a look?

File: utilities/write_batch_with_index/write_batch_with_index_internal.h (hunk 2 above)
Comment (cbi42) [reply utilities/write_batch_with_index/write_batch_with_index_internal.h:227]: @ywave620 I don't think this is addressed yet. See the following unit test:
```
TEST_P(WriteBatchWithIndexTest, CheckBounds) {
  Status s = OpenDB();
  ASSERT_OK(s);
  KVMap empty_map;
  // writes that should be observed by BaseDeltaIterator::delta_iterator_
  ASSERT_OK(batch_->Put("c", "cc"));
  ReadOptions ro;
  ro.iterate_lower_bound = new Slice("a");
  ro.iterate_upper_bound = new Slice("b");
  std::unique_ptr<Iterator> iter(batch_->NewIteratorWithBase(
      db_->DefaultColumnFamily(), new KVIter(&empty_map), &ro));
  iter->SeekToFirst();
// This assert fails
  ASSERT_FALSE(iter->Valid());
  delete ro.iterate_lower_bound;
  delete ro.iterate_upper_bound;
}
```

**Review by cbi42 — CHANGES_REQUESTED** [review]
https://github.com/facebook/rocksdb/pull/11680#discussion_r1342960097

**Comment by cbi42** [comment]
I think the additional bounds check also apply to Seek and SeekForPrev in addition to SeekTo{First,Last}.

**Comment by ywave620** [comment]
> I think the additional bounds check also apply to Seek and SeekForPrev in addition to SeekTo{First,Last}.

Also adopted, PTAL.

**Comment by adamretter** [comment]
@ywave620 May I ask if this will also cover the previous work done in https://github.com/facebook/rocksdb/pull/7214

**Comment by ywave620** [comment]
> @ywave620 May I ask if this will also cover the previous work done in #7214 please?

Yes. But this PR takes care more than that. And sorry, I didn't know it when I started this work.

**Comment by adamretter** [comment]
@ywave620 Thank you. I will just be happy if the problem is fixed and the code is merged :-)

**Review by cbi42 — APPROVED** [review]
LGTM, thanks!

**Comment by ywave620** [comment]
@cbi42 Thanks for your detail review. Is this ready to be merged? :)

**Comment by cbi42** [comment]
@ywave620 Could you also add a change log under https://github.com/facebook/rocksdb/tree/main/unreleased_history?

Besides that, it just needs internal review now.

**Comment by ywave620** [comment]
@cbi42 Hey, is there any news? :)

**Comment by ajkr** [comment]
> Besides that, it just needs internal review now.

> @cbi42 Hey, is there any news? :)

The internal reviewer is me, but I haven't gotten to it yet. Probably can do it tomorrow

File: utilities/write_batch_with_index/write_batch_with_index_internal.h [hunk 3]
```diff
@@ -293,6 +339,45 @@ class WBWIIteratorImpl : public WBWIIterator {
...
+  bool ValidRegardlessOfBoundLimit() const {
+    if (!skip_list_iter_.Valid()) {
+      return false;
+    }
+    const WriteBatchIndexEntry* iter_entry = skip_list_iter_.key();
+    return iter_entry != nullptr &&
+           iter_entry->column_family == column_family_id_;
+  }
+
+  bool AfterUpperBound(const Slice* k) const {
```
Comment (ajkr) [inline utilities/write_batch_with_index/write_batch_with_index_internal.h:360]: I'd call it `AtOrAfterUpperBound()`

**Review by ajkr — APPROVED** [review]
Great work, thanks!

**Comment by ywave620** [comment]
@ajkr The branch goes out-of-date once again, I can't keep it synchronous consistently, please update it before you are going to merge :)

**Pull request closed without merging**